### ISSUE TITLE
Commit message editor should help add "Co-authored-by: " text

### ISSUE THREAD

**Issue by drewnoakes** [gitextensions/gitextensions#5903]
**Do you want to request a *feature* or report a *bug*?**

Feature.

**What is the expected behavior?**

To be clear, I'm talking about the editor that pops up via `GitExtensions.exe editor` (I get it for `git commit` on the command line, for example).

When a commit has multiple authors, authors other than the committer may be specified by including a line such as:

```text
Co-authored-by: Alice Zachary <alice@foo.com>
```

While this can be typed manually, it's something I personally use infrequently and have to search for whenever I need it. It'd be a useful feature for me on those occasions.

The commit message editor has a toolbar with one button (save). We could add another icon there.

There may be other lines, such as `Signed-off-by:`.

These shouldn't be visible when editing interactive rebase TODO scripts. We already have logic to detect such cases (to control syntax highlighting).

**Environment you encounter the issue:**

- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc

**Comment by drewnoakes** [comment]
Actually this would be useful in `FormCommit` too.

**Comment by SeanFeldman** [comment]
I'd like to help with this as I need this feature often. Any recommendations / pointers where to start?

**Comment by vbjay** [comment]
@SeanFeldman create a new menu in this section that says Insert Co-authored-by.  It will create text like 

```
Co-authored-by: NAME <Email Address>
```
![image](https://user-images.githubusercontent.com/3629489/50711069-e6edc300-103a-11e9-8b80-43a77af1f08c.png)

Use our code to get the name and email address info and place it in the correct spot.  See how the text in the status bar of the commit dialog is created.

**Comment by vbjay** [comment]
Also.  Create an insertable template provider so adding other like things can be easily created and other UI components like the commit message editor can all use the same functionality.  Keep the code DRY :D

**Comment by vbjay** [comment]
Very rough idea and pseudo code
```
class CommitMessagePartialProvider{
virtual ToString()
{return ""}

}

class CoAuthoredMessageProvider : CommitMessagePartialProvider{
override ToString()
{return $"{getName()} <{getEmail()}>}"
}
....
```
Create a factory class that lets you create the right instance.  This way each class can know how to get the right data to generate the message.  Just like CoAuthored would need user name and email address.  Other templates would need other info.

**Comment by RussKie** [comment]
I'd like to avoid inheritance, if we can help it, and instead use composition.
In this case, we'll have a common interface and several implementations of it (for each use case), and a factory (as Jay suggested).

Something like:

```
public interface ICommitMessagePartialProvider
{
    string GetMessage();
}

public sealed class CoAuthoredMessageProvider : ICommitMessagePartialProvider
{
    public GetMessage()
    {
        return $"{getName()} <{getEmail()}>}";
    }
}

public sealed class CommitMessagePartialProviderFactory
{
    public ICommitMessagePartialProvider Create(CommitMessagePartialType type)
    {
        switch (type)
        {
            case CommitMessagePartialType.CoAuthored:
            {
                return new CoAuthoredMessageProvider();
            }
            default:
            {
                throw new NotSupportedException();
            }
        }
    }
}
```

Though before we jump straight into implementation can we talk through the UI and UX - how do you see the feature working?

**Comment by SeanFeldman** [comment]
@RussKie glad you brought that up. I'd rather focus in functionality and not worry about reuse. Reuse can cause abuse 😉

I haven't looked at the code yet to provide any valuable feedback.

**Comment by vbjay** [comment]
The reuse I was talking about was the fact that we have 
![image](https://user-images.githubusercontent.com/3629489/50720922-aadf5000-1084-11e9-9b08-f4863f1abe35.png) and 
![image](https://user-images.githubusercontent.com/3629489/50720943-0dd0e700-1085-11e9-8cef-8f19010c9b7e.png)  So putting the code to generate the message in the FormCommit is foolish.  Why I suggested a provider class.

**Comment by drewnoakes** [comment]
@SeanFeldman great to hear you'd like to pick this up. Why don't you start simple and get this working in one place. If you open a PR with your work as you do it, we can provide more concrete pointers.

For now, maybe add a separator and two new menu items to the commit menu (as @vbjay says). They could read "Sign off" and "Add co-author".

They should append any needed whitespace, then add the relevant text. For sign off, it'd make sense to append the user's name and email too.

You could create a reusable component that takes the current message text and returns the modified text. We could then use that component in both kinds of UI that edits messages. You can add some tests for that component too.

How does that sound?

**Comment by SeanFeldman** [comment]
Sorry for delay. I've actually found templates to be very helpful. I've defined one of the templates, as a `Co-authored` template and use it when pairing.

**Comment by mkphph** [comment]
I see this ticket hasn't been assigned and have been looking into it for some time. So far the comments here are useful enough to get me started I would pick it up if that's okay.

**Comment by RussKie** [comment]
Sounds good, thank you.

**Comment by pmiossec** [comment]
> I see this ticket hasn't been assigned and have been looking into it for some time. So far the comments here are useful enough to get me started I would pick it up if that's okay.

I don't know what you are planning to do but FYI, I have done something for the commit form some times ago to have the conventional commit message keywords available. But I have added also some footer keywords: 

https://github.com/pmiossec/gitextensions/commit/52effb55e60a38e3707febbf81877165054d8dad

I don't say that is the way to go but it could be useful to know to discuss about it...

The initial feature request is for the editor but I don't know if I'm representative of GitExtensions users or not but I nearly never use it and will see more added value by having the feature available in the commit form like discussed in later comments.

Thinking about it again, an interesting possibility could be to inject these key words into the autocompletion built word list so that when we start typing 'co' we will have the 'Co-authored-by:' string directly available.

**Comment by mkphph** [comment]
Hey, sure, let's discuss, I agree I wasn't clear on my intentions.

So I am new to contributing and to Git Extensions code so I planned to start with the suggestions above. Briefly:
1) Add option 'Co-authored by' in the FormCommit (ToolStripMenu). Just make it work, functional; get familiar with the basecode, be at ease playing with it.
2) Add option 'Sign off', same mindset as 1)
3) After the 2 options are functional, thinking bigger and try to implement through interface.
4) Cleanup
5) Ask for review (PR as draft)
6) Unit tests

I was just at step 2 when I realized the Sign-off option already exists (Options, at the far right). I guess this would cancel 2) and 3). In any case, very open to feedback here.

Does that make sense? and is the feature still wanted/relevant?

**Comment by vbjay** [comment]
Just throwing idea your way. Add a dialog that lets users give author name and email address as a kind of collection editor. Add option to launch dialog on commit message menu item. When dialog is submitted with the list of name/email pairs you append at bottom the generated items. 

Bonus feature...work in GitHub plugin to allow auto complete by typed user names so vbja would suggest my user name and my GitHub no reply address.

**Comment by drewnoakes** [comment]
Building on @vbjay's suggestion (which could totally be done separately, so don't get blocked on that), but I think it'd be enough to have an autocomplete of existing authors/committers in the repo. It can be tricky to get the name/email correct, so completion would be a real help. But again, just having something help add the correct co-author syntax would avoid me having to search online every time I use it :)

**Comment by RussKie** [comment]
> 2\. Add option 'Sign off', same mindset as 1)

We already have a functionality to do so (NB: I don't use it, so I can't comment on how usable it is):
![image](https://github.com/gitextensions/gitextensions/assets/4403806/7a22d152-8584-44c0-b689-f4247424f380)

It could be that we should add a popup option next to the textbox with few options:
* add own name and email as configured in git config (for the current repo)
* add a name and email of an existing contributor to the repo (run [`git shortlog`](https://git-scm.com/docs/git-shortlog))

...or

* open a dropdown list which contains the user's name+email as well as the list of all other contributors for a picking. E.g., similar to this:
![image](https://github.com/gitextensions/gitextensions/assets/4403806/3efc5377-bfc1-443f-aa8b-761b84e58153)


I'm not sure which UX would serve better, just thinking out loud.

**Comment by RussKie** [comment]
...or turn this textbox into a dropdown control which allows free text as well as picking an item from a list, if a user opens the dropdown portion of it.

**Comment by vbjay** [comment]
I for one think it shouldn't be in options drop-down. I would first look at the commit message button. Also launching a dialog that lets user enter name and email address and hit add. It adds entries to a list that are the generated values. The name text box autocompleting GitHub user names and potential for other sources of user names like gitlab and azure DevOps and... by selecting a source to be filled with the supported sources. 

Yes the auto complete is just a nice to have and not saying it's a deal breaker. Just saying oyes an extension of the idea. 

Another thought. If people know of other special tags besides co-authored that could be a selectable item in dialog. Co-authored by, tested by.... Whatever. Could be a combobox with text allowed to type and have predefined options to select.

**Comment by RussKie** [comment]
My dropdown suggestion was directed towards the "sign-off" option. For the "co-authored" option that won't work well because there could be multiple authors.

**Comment by mkphph** [comment]
> [pmiossec@52effb5](https://github.com/pmiossec/gitextensions/commit/52effb55e60a38e3707febbf81877165054d8dad)
I'll take the time to have a look.

> Bonus feature...work in GitHub plugin to allow auto complete by typed user names so vbja would suggest my user name and my GitHub no reply address.
I think this is next level for me at the moment lol but it's interesting.

Ok nice, lots of ideas. Trying to synthesise here, I think I would try with adding an item in the Commit message button. Sthg like "Add co-authors", "Add collaborators". You click the item, it opens a dialog box containing authors from the project (via git shortlog), and you check them to be added to a List, and the checked ones are displayed in the message. Maybe add a "Tester" or "Co-author" flag to it... Sounds good for a start?

**Comment by pmiossec** [comment]
> Ok will also have to learn how to make proper replies.

Just add an empty line in between.

**Comment by drewnoakes** [comment]
#11402 is awesome and definitely helps, but requires you to remember that the option you want starts with "Co-". I believe we should also add a command to the menu that inserts these standard strings a the end of the commit message, as those commands would be more discoverable. They could even launch a UI to select a known user from the repo.

**Comment by pmiossec** [comment]
> #11402 is awesome and definitely helps, but requires you to remember that the option you want starts with "Co-".

Not necessarily. If you open the completion (<kbd>Ctrl+space</kbd>) on a new line (or at least not on an existing word), you will only have the keywords added like that:
![image](https://github.com/gitextensions/gitextensions/assets/460196/6560f6a6-03f4-459e-bbfa-ba1eeaef7e83)

>I believe we should also add a command to the menu that inserts these standard strings a the end of the commit message, as those commands would be more discoverable.

I also have that in stock and will made a PR not far from now, now that I have a workflow that I find convenient...

![image](https://github.com/gitextensions/gitextensions/assets/460196/fdd451e4-0a11-4e6d-9999-1825b3840436)
